### PR TITLE
[WIP] Revert "Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)" and enable closed form with non-sequential epoch parameter

### PR DESCRIPTION
**Pull request by vincentqb**
We want to let the user combine schedulers, as described in #13022 and [here](https://github.com/facebookresearch/maskrcnn-benchmark/blob/master/maskrcnn_benchmark/solver/lr_scheduler.py#L7-L9). The mechanic of combining schedulers together implemented in #14010 results in unexpected behavior for the user, see #20527. We are planning on reinstating #14010 with the following modifications: when the user inputs an epoch value different from the upcoming one, a warning is shown, and the closed form of the scheduler is used when available.

This PR also defines the behavior of `get_lr` to return the last computed value of a scheduler to address #22107.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21800 Revert "Revert "Redefine scheduler to set learning rate using recursive formula" #14010 (#21463)" and enable closed form with non-sequential epoch parameter**

This reverts commit 3889855a5baf77cbdad60b94bcad77fdd7f1b145.

Differential Revision: [D15833938](https://our.internmc.facebook.com/intern/diff/D15833938)

### CODE REVIEW

**Comment by ezyang** [comment]
Since this isn't a "straight" revert, we should have a more detailed description in the PR body explaining what the approach we took in this version of the patch is. We should also mention which bug we fix.

File: test/test_optim.py
Comment (ezyang) [inline test/test_optim.py]: Great!

File: test/test_optim.py
Comment (ezyang) [inline test/test_optim.py]: I don't see why you need this ignore here. We're not stepping, right?

Comment (ezyang) [reply test/test_optim.py]: (I won't comment on the other occurrences of this.)

File: test/test_optim.py
Comment (ezyang) [inline test/test_optim.py]: Not just almost equal, exactly equal, right?!

File: torch/optim/lr_scheduler.py
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: I suppose `get_lr` isn't a publicly documented function, but it would be nice to explain in English, somewhere, what exactly `get_lr_closed_form` does. You could do a Note: somewhere in this file, write your comments below a header `Note [LR closed form]` and then reference this note elsewhere.

File: torch/optim/lr_scheduler.py
Comment (ezyang) [inline torch/optim/lr_scheduler.py]: This still needs to get fixed right?

Comment (ezyang) [reply torch/optim/lr_scheduler.py]: Actually, now I'm surprised the tests are passing. What happened here?

**Review by ezyang — COMMENTED** [review]
This looks good! Just a little more documentation (esp. in the PR body) needed, and I'll stamp it.

File: test/test_optim.py
Comment (vincentqb) [reply test/test_optim.py]: In the tests referred, __init__ is executing step with a nonzero initial step which triggers the warning.

File: test/test_optim.py
Comment (vincentqb) [reply test/test_optim.py]: The `step` operator will recompute the value at each iteration, and indeed the two real numbers should be equal, yes. I can replace by `assertEqual` if we are not worried about machine precision.

File: torch/optim/lr_scheduler.py
Comment (vincentqb) [reply torch/optim/lr_scheduler.py]: The issue was that the learning rate around line 84 was computed before the epoch was incremented. That has been fixed above.

File: test/test_optim.py
Comment (vincentqb) [reply test/test_optim.py]: The tests pass with `assertEqual`, so I've updated the tests.

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
@fmassa could you take a look too?

**Comment by ezyang** [comment]
@vincentqb Before we land this we should also make sure that #22107 is also fixed.

**Comment by ezyang** [comment]
It looks like it still has this bug:

```
(/scratch/ezyang/pytorch-build-env) ezyang@devfair040:/scratch/ezyang/pytorch-build$ python test.py
0 [0.1]
/scratch/ezyang/pytorch-build/torch/optim/lr_scheduler.py:87: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
1 [0.1]
2 [0.1]
3 [0.0010000000000000002]
4 [0.010000000000000002]
5 [0.010000000000000002]
6 [0.00010000000000000003]
7 [0.0010000000000000002]
8 [0.0010000000000000002]
9 [1.0000000000000004e-05]
```

**Comment by vincentqb** [comment]
The recursive behavior means that calling `get_lr` will compute the new learning rate based on the current learning rate (which has changed when using step). `get_lr` doesn't set the learning rate and so the training is not affected. The code behave as it should, but this can be confusing to the user, as in  #22107.

We can move the computation logic away from `get_lr` to a private function, and make it so that `get_lr` returns either `params['lr']` independent of the scheduler, or returned the last computed learning rate by the current scheduler (with potentially a user warning, and at the small cost of having to save the most recent learning rate for each scheduler). I expect the latter to be the behavior most expected by the user given #22107. Thoughts?

**Comment by vincentqb** [comment]
The recursive behavior means that calling `get_lr` will compute the new learning rate based on the current learning rate (which has changed when using step). `get_lr` doesn't set the learning rate and so the training is not affected. The recursive code behave as it should, but this can be confusing to the user, as in  #22107.

We can move the computation logic away from `get_lr` to a private function, and make it so that `get_lr` returns either `params['lr']` independent of the scheduler, or returned the last computed learning rate by the current scheduler (with potentially an user warning, and at the small cost of having to save the most recent learning rate for each scheduler). I expect the latter to be the behavior most expected by the user, given #22107. Thoughts @ezyang?

**Comment by ezyang** [comment]
That sounds plausible to me. Maybe we can express this most clearly by writing some updated documentation about what exactly LRScheduler semantics are and asking users if they agree with this?

**Comment by vincentqb** [comment]
Here's to document the semantics discussed.

The current learning rate can be access through `optimizer.param_groups` as below:

```
import torch
from torch.nn import Parameter

optimizer = torch.optim.SGD([Parameter(torch.randn(2, 2, requires_grad=True))], 0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)

for i in range(10):
  print(i, [group['lr'] for group in optimizer.param_groups])  
  scheduler.step()
```

Since `get_lr` is [not a publicly documented function](https://github.com/pytorch/pytorch/pull/21800#discussion_r294305485), we can make `get_lr` return the last computed learning rate by the given scheduler, with a deprecation warning. This would look like what was entered here:

```
for i in range(10):
  print(i, scheduler.get_lr())  # Deprecation warning
  scheduler.step()
```

We could also make `step` return its computed value. This would make it clear that the algorithm is also stepping forward. This would look like:

```
for i in range(10):
  print(i, scheduler.step())
```

**Comment by vincentqb** [comment]
Based on [feedback](https://github.com/pytorch/pytorch/issues/22107#issuecomment-506260305), there is a preference for `get_lr` returning the most recent value computed by the scheduler as the preferred mechanism. I've updated the PR to reflect that.

**Comment by vincentqb** [comment]
Bug #21965 might be related to recursive schedulers.

File: test/test_optim.py
Comment (accelsao) [inline test/test_optim.py]: why [0.00005] * 3 instead of [0.00005] * 2 ?

Comment (vincentqb) [reply test/test_optim.py]: The test checks at each one of the 10 iterations, so the test still pass with a longer target list. I'll still make it `[0.00005] * 1` for clarity.

**Comment by vincentqb** [comment]
@fmassa -- Here are possible implementations for chainable schedulers:
1. The one implemented in this pull request uses by default the chainable formula, but when a user calls with a different epoch the closed form is used when available.
2. We can leave the current "closed form" schedulers untouched, and create chainable versions of those that can be chained, with a name like ChainableLRScheduler.
3. Each scheduler that can be chained has a chainable option when created.
4. A wrapper takes a sequence of scheduler, and chains them.

Based on [feedback](https://github.com/pytorch/pytorch/issues/22107#issuecomment-506260305), I'm going for (2 -- ChainableLRScheduler) to keep backward compatibility when users use current schedulers. Thoughts?

**Comment by vincentqb** [comment]
#22798 implements the change in the previous comment on top of this one.

**Review by ezyang — CHANGES_REQUESTED** [review]
Needs to be squashed with the subsequent PR, please do not merge this one alone.

**Comment by vincentqb** [comment]
> @fmassa -- Here are possible implementations for chainable schedulers:
> 
> 1. The one implemented in this pull request uses by default the chainable formula, but when a user calls with a different epoch the closed form is used when available.
> 2. We can leave the current "closed form" schedulers untouched, and create chainable versions of those that can be chained, with a name like ChainableLRScheduler.
> 3. Each scheduler that can be chained has a chainable option when created.
> 4. A wrapper takes a sequence of scheduler, and chains them.
> 
> Based on [feedback](https://github.com/pytorch/pytorch/issues/22107#issuecomment-506260305), I'm going for (2 -- ChainableLRScheduler) to keep backward compatibility when users use current schedulers. Thoughts?

Here's some feedback based on a discussion with @fmassa:

**About (1)**

The initial issue with #14010 (reverted in #21463) was #20527. However, #20527 was about unexpected behavior when passing the epoch parameter (for instance, when reading the learning rate -- a use case which is addressed here by documenting and providing `get_lr` method). If we deprecate the epoch parameter, that is no longer an issue, and step can chain by default. Let's note that the epoch parameter does not make sense for all schedulers even today (e.g. ReduceLROnPlateau).

The [user feedback mentioned](https://github.com/pytorch/pytorch/issues/22107#issuecomment-506260305) warns against unexpected result when changing the behavior of schedulers. However, the pull request it lives in (#22107) was about reading the current learning rate through a function that is not officially documented which was in fact used for computation. This is also address in the pull request by officially sanctioning `get_lr` to return its last computed learning rate.

**About (2)**

Having two separate classes might be confusing to the user (e.g. which version to use when?) duplicates code, or adds complexity in the code (e.g. by providing a scheduler factory). Moreover, some logic that differs between the two classes might end up living in the other, say as in the current version of #22798. 

**About (3)**

This is probably simpler than (2) to implement. However, this still add complexity to the code when only one version would be needed.

**About (4)**

What happens when two Wrapper are present? The current implementation simply chains everything. This may be confusing to the user. 

One solution is to have some sort of global lock, but this may prove to rigid in the future. Moreover, this adds interaction and complexity when invoking many schedulers.

**Comment by vincentqb** [comment]
@fmassa @ezyang @luzhilin19951120

Based on the feedback, I propose to go with:

* Changing the behavior of schedulers to the chainable formula
* Using the closed form whenever epoch is different from None until the next release with a deprecation warning
* Removing the epoch parameter after deprecation
* Making `get_lr` the supported way of obtaining the last computed learning rate by the scheduler

To repeat the previous comment, this has the benefits of:
* Seamless transitioning to chainable schedulers
* Simplifying the scheduler logic, and avoiding code duplication
* Removing the epoch parameter, which has different behavior across schedulers
* Having a consistent documented way of getting the last learning rate computed by a scheduler

Thoughts?

**Comment by vincentqb** [comment]
Further discussion with @fmassa.

The chaining of schedulers touches of course on the core question of what do we want schedulers to be. An [early implementation](https://github.com/pytorch/examples/blob/013f9573eda300484cdfb429a0424ab54e0014f1/imagenet/main.py#L395-L399) was a simple function. #1370 then mimicked keras in the implementation of schedulers. We are now discussing abstracting the concept of scheduler, see #22343, to enable other quantities (note: `get_lr` should in fact be named `get_last_values` to be more generic). However, we want to make sure that users realize it is easy to implement their own schedulers.

**Comment by vincentqb** [comment]
I had a discussion with @fmassa to consider replacing the scheduler's internal mechanic to be that each scheduler provides only a multiplicative factor to some abstract factory that knows how to apply it to the learning rate of each parameter. However, it appears not all scheduler are easily generated through such multiplicative factor, `CosineAnnealingLR` is probably one such.

**Comment by vincentqb** [comment]
Based on the feedback, I propose to go with:

* Changing the behavior of schedulers to the chainable formula
* Using the closed form whenever epoch is different from `None` until the next release with a deprecation warning
* Removing the epoch parameter and the closed form after deprecation
* Making `get_computed_values` the supported way of obtaining the last computed learning rate by the scheduler
* Returning a deprecation warning when invoking the undocumented `get_lr` function (see [comment](https://github.com/pytorch/pytorch/pull/21800#issuecomment-506051090)) referring to `get_computed_values`, and deprecating it in the next release.

This has the benefit of:

* Providing a natural syntax for composition of schedulers
* Seamlessly transitioning to chainable schedulers and maintaining backward compatibility with current syntax
* Keeping the scheduler logic leaner
* Removing the epoch parameter, which has different behavior across schedulers
* Having a consistent documented way of getting the last learning rate computed by a scheduler

@fmassa @ezyang @soumith @luzhilin19951120 -- thoughts?

**Comment by vincentqb** [comment]
> Needs to be squashed with the subsequent PR, please do not merge this one alone.

#21800 will be self-sufficient moving forward, since #22798 has been close since we are going in a different direction.

**Comment by vincentqb** [comment]
By the way, this would be the new way of address the two related PR.

#20527

```
import torch.optim as optim
from torch import nn

conv = nn.Conv2d(3,3,3)
optimizer = optim.Adam(conv.parameters()) 
lr_scheduler = optim.lr_scheduler.StepLR(optimizer, 2)

for _ in range(10):
    # NOTE Remove epoch parameter
    lr_scheduler.step()
    print(optimizer.param_groups[0]['lr'])
```

#22107

```
import torch
from torchvision.models import resnet18
net = resnet18()
optimizer = torch.optim.SGD(net.parameters(), 0.1)
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones=[3, 6, 9], gamma=0.1)
scheduler = torch.optim.lr_scheduler.StepLR(optimizer, 3, gamma=0.1)
for i in range(10):
    # NOTE Invoke new interface to get values
    print(i, lr_scheduler. get_computed_values())
    scheduler.step()
```

**Comment by vincentqb** [comment]
Discussion with @ezyang: If someone wants to use a closed form, they can always write a simple function setting `optimizer.param_groups[i]['lr']`. When using a closed form, a user shouldn't expect to chain schedulers.

**Comment by vincentqb** [comment]
`CosineAnnealingWarmRestarts`
* uses an old pattern to run with the optimizer, see [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L672)
* currently requires the epoch parameter in `step` since i doesn't keep track of fractional stepping.

#23306 implements the changes mentioned [here](https://github.com/pytorch/pytorch/pull/21800#issuecomment-513370208), and keeps `CosineAnnealingWarmRestarts` as it currently is.

**Comment by SebastienEske** [comment]
Hello,

I noticed that at the end of the scheduler initialization the step function is called
`self.step(last_epoch)`
line 44 of https://github.com/pytorch/pytorch/blob/364d24cb65e4f35358911a208a8cae0cd0f8431a/torch/optim/lr_scheduler.py

This results in applying immediately the learning rate modification for the recursive schedulers such as StepLR while it should happen at the next iteration/epoch. If needed I can log another bug for this. In any case, I believe it should be fixed. What do you think?

**Comment by vincentqb** [comment]
Superseded by #23306.

**Pull request closed without merging**
